lib/scholar/cluster/gmm.ex

```diff
@@ -0,0 +1,433 @@
+                    converged = Nx.tensor(false),
+                    x,
+                    key,
+                    max_iter,
+                    tol,
+                    covariance_regularization_eps
+                  }
+                },
+                iter <= max_iter and not converged do
+            prev_lower_bound = lower_bound
+            {weights, means, _, precisions_cholesky} = params
+            {log_prob_norm, log_responsibilities} = e_step(x, weights, means, precisions_cholesky)
+            params = m_step(x, log_responsibilities, covariance_regularization_eps)
+            lower_bound = log_prob_norm
+            change = lower_bound - prev_lower_bound
+            converged = iter > 1 and Nx.abs(change) < tol
+
+            {
+              {lower_bound, params},
+              {iter + 1, converged, x, key, max_iter, tol, covariance_regularization_eps}
+            }
+          end
+
+        {max_lower_bound, best_params} =
+          if max_lower_bound < lower_bound do
+            {lower_bound, params}
+          else
+            {max_lower_bound, best_params}
+          end
+
+        {
+          best_params,
+          {run + 1, max_lower_bound, x, key, max_iter, tol, covariance_regularization_eps}
+        }
+      end
+
+    {weights, means, covariances, precisions_cholesky} = best_params
+
+    %__MODULE__{
+      weights: weights,
+      means: means,
+      covariances: covariances,
+      precisions_cholesky: precisions_cholesky
+    }
+  end
+
+  defnp init_params(key, x, opts) do
+    type = Nx.type(x)
+    {num_samples, _} = Nx.shape(x)
+    num_gaussians = opts[:num_gaussians]
+    covariance_regularization_eps = opts[:covariance_regularization_eps]
+
+    k_means_model =
+      Scholar.Cluster.KMeans.fit(x, num_clusters: num_gaussians, num_runs: 1, key: key)
+
+    labels = k_means_model.labels
+    responsibilities = Nx.take(Nx.eye(num_gaussians, type: type), labels)
+
+    {weights, means, covariances, precisions_cholesky} =
+      estimate_gaussian_parameters(x, responsibilities, covariance_regularization_eps)
+
+    weights = weights / num_samples
+    {weights, means, covariances, precisions_cholesky}
+  end
+
+  defnp e_step(x, weights, means, precisions_cholesky) do
+    {log_prob_norm, log_responsibilities} =
+      estimate_log_prob_responsibilities(x, weights, means, precisions_cholesky)
+
+    {Nx.mean(log_prob_norm), log_responsibilities}
+  end
+
+  defnp m_step(x, log_responsibilities, covariance_regularization_eps) do
+    {weights, means, covariances, precisions_cholesky} =
+      estimate_gaussian_parameters(x, Nx.exp(log_responsibilities), covariance_regularization_eps)
+
+    weights = weights / Nx.sum(weights)
+    {weights, means, covariances, precisions_cholesky}
+  end
+
+  defnp estimate_gaussian_parameters(x, responsibilities, covariance_regularization_eps) do
+    nk = Nx.sum(responsibilities, axes: [0])
+    means = Nx.dot(Nx.transpose(responsibilities), x / nk)
```

Comment:
```suggestion
    means = Nx.dot(responsibilities, [0], x / nk, [0])
```

I believe this will achieve the correct result too, but avoiding the transpose.

```elixir
iex(37)> k = Nx.Random.key(System.system_time())
#Nx.Tensor<
  u32[2]
  EXLA.Backend<host:0, 0.3654900815.2985164818.202632>
  [393393791, 2867920131]
>
iex(38)> {t1, k} = Nx.Random.uniform(k, shape: {3, 4})
{#Nx.Tensor<
   f32[3][4]
   EXLA.Backend<host:0, 0.3654900815.2985164818.203175>
   [
     [0.13576042652130127, 0.8412826061248779, 0.4484595060348511, 0.5624940395355225],
     [0.3512943983078003, 0.004010677337646484, 0.32932519912719727, 0.3748883008956909],
     [0.5318529605865479, 0.8757333755493164, 0.39265263080596924, 0.1744011640548706]
   ]
 >,
 #Nx.Tensor<
   u32[2]
   EXLA.Backend<host:0, 0.3654900815.2985164818.203177>
   [1882076740, 466335474]
 >}
iex(39)> {t2, k} = Nx.Random.uniform(k, shape: {3, 5})
{#Nx.Tensor<
   f32[3][5]
   EXLA.Backend<host:0, 0.3654900815.2985164818.203698>
   [
     [0.36859428882598877, 0.4500901699066162, 0.6306818723678589, 0.5150032043457031, 0.25934934616088867],
     [0.5741105079650879, 0.45720410346984863, 0.05595088005065918, 0.6362648010253906, 0.32299208641052246],
     [0.1018531322479248, 0.21718335151672363, 0.5860718488693237, 0.7845090627670288, 0.2448132038116455]
   ]
 >,
 #Nx.Tensor<
   u32[2]
   EXLA.Backend<host:0, 0.3654900815.2985164818.203700>
   [1307000727, 4256395817]
 >}
iex(40)> Nx.equal(Nx.dot(Nx.transpose(t1), t2), Nx.dot(t1, [-2], t2, [-2]))       #Nx.Tensor<
  u8[4][5]
  EXLA.Backend<host:0, 0.3654900815.2985164818.203708>
  [
    [1, 1, 1, 1, 1],
    [1, 1, 1, 1, 1],
    [1, 1, 1, 1, 1],
    [1, 1, 1, 1, 1]
  ]
>
```

Reply:
How about `means = Nx.dot(responsibilities, [0], x / nk, [0])`?
Using `[0]` instead of `[-2]` seems more understandable to me.

Reply:
I went with -2 due to force of habit of supporting batched versions of algorithms haha

Both are fine to me if you're enforcing 2D tensors here

Reply:
Yeah, I had a similar habit with Tensorflow / PyTorch. I would use `[0]` here as both are matrices.